internal/log_analysis/rules_engine/src/rule.py

```
@@ -81,12 +87,14 @@ def run(self, event: Dict[str, Any]) -> RuleResult:
             rule_result = _run_command(self._module.rule, event, bool)
             if rule_result and self._has_dedup:
                 dedup_string = _run_command(self._module.dedup, event, str)
+                if len(dedup_string) > MAX_DEDUP_STRING_SIZE:  # type: ignore
+                    dedup_string = dedup_string[:MAX_DEDUP_STRING_SIZE]  # type: ignore
```

Any reason not to fail hard? This silently changes behavior. I would prefer a fail here.

---

If we fail, this event will not be considered as failing the rule (we run the dedup function only for events that matched a rule). It will never pop up in an alert. 

I think it is more safe behavior to just truncate it, so the event will be grouped with different set of events. The limit is pretty high in any case 1000 characters should be enough for most dedup scenarios. 

I plan to make sure it is clear the behavior to the users as well:
- Explain the behavior in the documentation
- When someone creates a new rule, we show a "sample" rule in UI. I plan to have a comment in the dedup function explaining the behavior.

---

> Any reason not to fail hard? This silently changes behavior. I would prefer a fail here.

I think the API should just reject dedup strings that are too long, you said it's provided by the user, right?

---

My understanding is that the dedup string can be dynamically generated based on different properties of the policy/resource, so it is hard to know how long it will be before it runs. I'd be okay with truncating and maybe a warning message or something? Truncating is better than just dropping it or not doing any deduping.

---

I'm very worried about strange things happening here, ask Nick pointed out the dedupe can be dynamic, otherwise we could have API check it (and fail hard here as a precaution, but it should never happen). Given it is dynamic, we are opening up to strange things. The worse that happens is more deduping is done than is desired but that will be super hard to debug. Perhaps we add an output field that records the fact that this event had a truncated dedupe string. Thoughts?

---

After understanding it better, truncating seems reasonable, but any reason we can't go with 5k or 10k as the limit?

---

> The worse that happens is more deduping is done than is desired but that will be super hard to debug. Perhaps we add an output field that records the fact that this event had a truncated dedupe string. Thoughts?

Sure, I will add some logging

---

> After understanding it better, truncating seems reasonable, but any reason we can't go with 5k or 10k as the limit?

1K is a mostly arbitrary limit. Some thoughts I had about this: 
- 400kb is the maximum size of a DDB item. We definitely have to be smaller than that. 
- 1 DDB write request unit is 1KB. Performing 1 write in DDB with 2KB payload will consume (and cost) 2 "writes". If some customer ends up really using 1k characters for deduping, given how we write to DDB this is going to cost him 3 DDB write units. 
- I initially had a limit of 256 to make sure a request to DDB is never more than 1 write request unit for cost and performance reasons. This would cover a lot of usecases e.g. max file name size in linux is 255 characters, maxi domain name size is 253, etc. I thought however to extend it more since maybe some customers might want to concatenate such fields. Most systems that I've seen doing deduping had much smaller limit so I thought that would be enough.
- I consider 1k characters to be enough for deduping for the vast majority of the use cases. Deduping string would be a specific log field (hostname, ip, username, region, ....) or maybe a combination of those. 1k characters seems already too large and probably it is a misuse of the field.
 - If a customer wants to dedup on something larger, it will actually be more efficient for them to just use as dedup a hash (MD5, SHA1, ...) of the payload than putting the full dedup payload. It is going to be cheaper on their bill as well. The limit is there also for the protection of the customer in case of potential misuse.  

These are the thoughts that went through my mind more or less. 
If you think it's not sufficient for some usecases, I could raise it further.